### DISCUSSION TITLE
Does the Hubitat integration still work?

### DISCUSSION THREAD

**Discussion by davthomaspilot** [tagyoureit/nodejs-poolController#458] [Q&A] [answered]
I see these messages in the Hubitat log:

`[sys:1](http://192.168.0.213/logs#sys1)2022-05-23 06:18:39.154 pm Received data from 192.168.0.31, no matching device found for 192.168.0.31, C0A8001F:BBF0, B827EB8AB26B or C0A8001F`

192.168.0.31 is the rpi running nodejs-poolController 192.168.0.213 is the Hubitat.

When I start a new PoolController app, I try discovering, but nothing is found.  I get these messages:

```
(http://192.168.0.213/installedapp/configure/291)SENT: lan discovery urn:schemas-upnp-org:device:PoolController:1
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.197 pm [debug](http://192.168.0.213/installedapp/configure/291)RESET DEVICES
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.194 pm [debug](http://192.168.0.213/installedapp/configure/291)RESET AS DEVICES
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.191 pm [debug](http://192.168.0.213/installedapp/configure/291)RESET DEVICES
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.188 pm [debug](http://192.168.0.213/installedapp/configure/291)RESET AS DEVICES
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.185 pm [debug](http://192.168.0.213/installedapp/configure/291)RESET DEVICES
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.183 pm [debug](http://192.168.0.213/installedapp/configure/291)RESET AS DEVICES
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:23:41.179 pm [debug](http://192.168.0.213/installedapp/configure/291)SENT: lan discovery urn:schemas-upnp-org:device:PoolController:1
```
```

```
When I enter the IP address for poolController manually, I get these messages:

```
rg:device:PoolController:1, hub:, port:4200, name:, ssdpPath:/device, ssdpUSN:uuid:806f52f4-1f35-4e33-9299-b827eb3bb77a::urn:schemas-upnp-org:device:PoolController:1, ssdpNTS:]]
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:26:39.186 pm [debug](http://192.168.0.213/installedapp/configure/291)DevDescHandler - > null
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:26:39.162 pm [debug](http://192.168.0.213/installedapp/configure/291)SENT: [:]
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:26:39.155 pm [info](http://192.168.0.213/installedapp/configure/291)Verify UPNP PoolController Device  b8:27:eb:8a:b2:6b @ http://192.168.0.31:4200/device
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:26:39.151 pm [debug](http://192.168.0.213/installedapp/configure/291)SENT: lan discovery urn:schemas-upnp-org:device:PoolController:1
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:26:37.209 pm [error](http://192.168.0.213/installedapp/configure/291)Cannot verify Device - No body in Device Response
[app:291](http://192.168.0.213/logs#app291)2022-05-23 06:26:37.206 pm [debug](http://192.168.0.213/installedapp/configure/291)DevDescHandler -devs- > [ b827eb8ab26b:[ipaddress:192.168.0.31, verified:false, uuid:, mac: b8:27:eb:8a:b2:6b, mode:, ssdpTerm:urn:schemas-upnp-org:device:PoolController:1, hub:, port:4200, name:, ssdpPath:/device, ssdpUSN:uuid:806f52f4-1f35-4e33-9299-b827eb3bb77a::urn:schemas-upnp-org:device:PoolController:1, ssdpNTS:]]
```

/sys/var/syslog on the poolController rpi has records like this:

`May 23 23:28:12 rem node[524]: [5/23/2022, 11:28:12 PM] #033[32minfo#033[39m: [1                                                                      1:28:12 PM] 192.168.0.213 GET /device {}`

I can do gets using postman using the "Swagger" api for api's like get http://192168.0.31:4200/state/all.   But, when I try a GET  for /device I get:

```
<!DOCTYPE html>
<html lang="en">

<head>
	<meta charset="utf-8">
	<title>Error</title>
</head>

<body>
	<pre>Cannot GET /device</pre>
</body>

</html>
```

So, I'm wondering if something in the api has changed that has messed up the Hubitat integration?  Or am I missing something?

**Comment by rstrouse** [comment, marked as answer]
I don't run Hubitat but the above has an invalid upnp request.  And yes this was previously in the njsPC code but it was not properly responding to upnp requests.

`urn:schemas-upnp-org:device:PoolController:1` is not the correct upnp schema.  The proper schema is `urn:schemas-tagyoureit-org:device:PoolController:1`

I think the original code was produced by @bsileo and is no longer supported but I submit a request to have him change that single line of code to recognize the correct schema.

**Reply by davthomaspilot** [reply]
thanks!

**Reply by bsileo** [reply]
I'm still around and still using this code. Saw the other request and will make that update as soon as I get a chance.

**Reply by davthomaspilot** [reply]
"is no longer supported "    Is there another integration for Hubitat that IS supported?

**Reply by davthomaspilot** [reply]
Brad, I tried changing  urn:schemas-upnp-org:device:PoolController:1 to urn:schemas-tagyoureit-org:device:PoolController:1 several places in the poolController App.

Strangely, a controller on 169.254.225.184:4200 was discovered (my local lan is 192.168.0.x and the controller is on .31

And now the pool controller logs this;

```
[app:325](http://192.168.0.213/logs#app325)2022-05-23 07:28:52.190 pm [debug](http://192.168.0.213/installedapp/configure/325)SENT: [:]
[app:325](http://192.168.0.213/logs#app325)2022-05-23 07:28:52.187 pm [info](http://192.168.0.213/installedapp/configure/325)Verify UPNP PoolController Device null @ http://169.254.225.184:4200/upnp.xml
[app:325](http://192.168.0.213/logs#app325)2022-05-23 07:28:52.183 pm [debug](http://192.168.0.213/installedapp/configure/325)SENT: lan discovery urn:schemas-tagyoureit-org:device:PoolController:1
```

**Reply by bsileo** [reply]
> "is no longer supported " Is there another integration for Hubitat that IS supported?

Well I would say this is still supported to the extent that any free open source code is! I am still using it in production daily running my system and able to make updates to the code!

**Reply by davthomaspilot** [reply]
Sounds good.  Appreciate it!

**Reply by bsileo** [reply]
> Brad, I tried changing urn:schemas-upnp-org:device:PoolController:1 to urn:schemas-tagyoureit-org:device:PoolController:1 several places in the poolController App.
> 
> Strangely, a controller on 169.254.225.184:4200 was discovered (my local lan is 192.168.0.x and the controller is on .31
> 
> And now the pool controller logs this;
> 
> ```
> [app:325](http://192.168.0.213/logs#app325)2022-05-23 07:28:52.190 pm [debug](http://192.168.0.213/installedapp/configure/325)SENT: [:]
> [app:325](http://192.168.0.213/logs#app325)2022-05-23 07:28:52.187 pm [info](http://192.168.0.213/installedapp/configure/325)Verify UPNP PoolController Device null @ http://169.254.225.184:4200/upnp.xml
> [app:325](http://192.168.0.213/logs#app325)2022-05-23 07:28:52.183 pm [debug](http://192.168.0.213/installedapp/configure/325)SENT: lan discovery urn:schemas-tagyoureit-org:device:PoolController:1
> ```

I just pushed new code for this change but have not tested it yet.  I have to go through updating my NJSPC to get to the latest.  Pool season is just starting here so I have not been working on that much recently :-)  I'll try to get a chance sometime this week and see what I find and see if I can reproduce your problem.

**Reply by davthomaspilot** [reply]
Thanks, Brad.

I just went through that too.  Then I needed a newer nodejs (or was it npm).  Then, that newer version complained it wasn't compatible with my old hardware (RPI 2B), so moved up to RPI 3B.  

Now everything working on nodejs-poolController and dashPanel.  I have Alexa bindings through Hubitat and your integration, so trying to get that going now.

Should I give what's up there now a try, or wait until later this week?

**Reply by davthomaspilot** [reply]
I gave it a try--no difference that I can tell.

**Reply by rstrouse** [reply]
@bsileo I think that line 100 needs to use the returned device url or hardcoded from
`ssdpPath:"/device",` to `ssdpPath:"/upnp.xml",`

**Reply by tagyoureit** [reply]
@rstrouse is spot on.  I changed it in my local code and everything syncs just fine.

**Reply by davthomaspilot** [reply]
Still not working for me:
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.382 pm [error](http://192.168.0.213/installedapp/configure/328)Cannot verify Device - No body in Device Response
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.379 pm [debug](http://192.168.0.213/installedapp/configure/328)DevDescHandler -devs- > [null:[deviceAddress:1068, ssdpTerm:urn:schemas-tagyoureit-org:device:PoolController:1, hub:1, verified:false, ssdpPath:/upnp.xml, ssdpUSN:uuid:806f52f4-1f35-4e33-9299-b827eb8ab26b::urn:schemas-tagyoureit-org:device:PoolController:1, networkAddress:A9FEE1B8, mac:null, ssdpNTS:null], b827eb8ab26b:[ipaddress:192.168.0.31, verified:false, uuid:, mac:b8:27:eb:8a:b2:6b, mode:, ssdpTerm:urn:schemas-tagyoureit-org:device:PoolController:1, hub:, port:4200, name:, ssdpPath:/device/unp.xml, ssdpUSN:uuid:806f52f4-1f35-4e33-9299-b827eb3bb77a::urn:schemas-tagyoureit-org:device:PoolController:1, ssdpNTS:]]
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.376 pm [debug](http://192.168.0.213/installedapp/configure/328)DevDescHandler - > null
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.348 pm [debug](http://192.168.0.213/installedapp/configure/328)SENT: [:]
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.338 pm [info](http://192.168.0.213/installedapp/configure/328)Verify UPNP PoolController Device b8:27:eb:8a:b2:6b @ http://192.168.0.31:4200/device/unp.xml
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.328 pm [debug](http://192.168.0.213/installedapp/configure/328)SENT: [:]
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.324 pm [warn](http://192.168.0.213/installedapp/configure/328)connect timed out
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.306 pm [info](http://192.168.0.213/installedapp/configure/328)Verify UPNP PoolController Device null @ http://169.254.225.184:4200/upnp.xml
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:44.302 pm [debug](http://192.168.0.213/installedapp/configure/328)SENT: lan discovery urn:schemas-tagyoureit-org:device:PoolController:1
[app:328](http://192.168.0.213/logs#app328)2022-05-24 03:09:42.396 pm [error](http://192.168.0.213/installedapp/configure/328)Cannot verify Device - No body in Device Response

**Reply by davthomaspilot** [reply]
Sorry, I mistyped line 100.

Now I think it's working!

Thanks bsileo!

**Reply by bsileo** [reply]
Ok I made this change in the repo.  Its  a little odd that this helped though as I thought that code would only matter in the case where you are manually adding your controller...when it is discovered with UPNP it should be getting this path dynamically already.  Been pretty busy recently so its going to be a bit till I get a chance to update my local environment and work through this.

**Reply by davthomaspilot** [reply]
@bsileo 

I manually entered the controller.

I couldn't get discovery to work.  A 169.x.x.x address was found, but not my controller.

Unless it takes more than a couple a minutes--I did the manual add when it didn't seem like it was going to work.  How long should I wait?

**Comment by rstrouse** [comment]
Thanks @bsileo!

**Reply by mhmatttu** [reply]
@bsileo per your suggestion, i made the switch to hubitat from smartthings.  certainly like the responsiveness of local control, but still figuring it all out.  i loaded your njsPC integration, but have no heater control.  i've tried all the templates in dashboard including thermostat.  i can read all the attributes for heater but can't control.  all works fine in dashpanel. also i've noticed device states in the hubitat ios app do not sync correctly.

**Reply by bsileo** [reply]
> @bsileo per your suggestion, i made the switch to hubitat from smartthings. certainly like the responsiveness of local control, but still figuring it all out. i loaded your njsPC integration, but have no heater control. i've tried all the templates in dashboard including thermostat.  i can read all the attributes for heater but can't control. all works fine in dashpanel.

What device are you looking at for heater control? This will be on a body, such as the Pool Body or Spa Body. 

![image](https://user-images.githubusercontent.com/6116016/170515349-15fee427-18ae-431c-89ec-a0fd37f791ab.png)

>  also i've noticed device states in the hubitat ios app do not sync correctly.

I am not sure about this one.  When you see this problem, are the states correct in the Hubitat website?  I do not use the iOS app. I use a platform called Sharptools.io for Dashboard and mobile access.

**Reply by davthomaspilot** [reply]
Brad,

How does one set currentHeaterMode?

I have both a heatpump (Solar) and gas heater.  In screenLogic I have choices:

OFF
Heater
H. Pump Pref  (no clue how it really works)
H. Pump.

I use H.Pump when not in a hurry to get it hot (or need to cool) since propane (Heater) is expensive.

Not a big deal (for m)  if it just stays on the previously selected mode after turning on then off.  I just want to turn on and off using Alexa--that's my main motivation for using the Hubitat integration.

**Reply by bsileo** [reply]
You can use Next Heater Mode to toggle ahead in the list.  But you are correct there is not a simple toggle to just set a particular mode.  Once the mode is set you can just turn it on and off as you noted, which I think works for most people. The one challenge is that the available Heating Modes is different for each installation, so the design was trying to handle that dynamic list.  I actually think I found some information on a way to create a dynamic set of options in Hubitat which if I could get to work would allow a "SetHeatingMode" command with correct options. Of course I could do it without options but then the user would have to type in the right value.

**Reply by mhmatttu** [reply]
> What device are you looking at for heater control? This will be on a body, such as the Pool Body or Spa Body.

Yes, I'm looking at pool and/or spa body.  From the device page as you showed, the setpoint commands and heater off command works but no response with heater on or next heater mode.

Within hubitat dashboard, the only templates that seem to work are attributes and temperature but these only show the data or attribute chosen, no command options.  I've read with other pool integrations, the thermostat template can be used for the heater.

> When you see this problem, are the states correct in the Hubitat website?

States are correct in the hubitat website and with hubitat dashboard, just not the hubitat ios app device page.  However, it seems many other devices have this issue...not just pool controller.  This is unfortunate as i would like to quickly view status or turn on/off a circuit without having to build a dashboard for everything.

I use (or used) Actiontiles with smartthings.  I was hoping to transition away from actiontiles and use hubitat dashboard to have all local control and not be cloud dependent, but unfortunately many options are missing from hubitat dashboard and design is difficult (in my opinion).

**Reply by davthomaspilot** [reply]
bsileo

Next Heater Mode is fine and on/off control is fine for what I need.  Whenever "Heat On" is pushed, the currentHeater Mode goes to Heater.  

So, I'll add a little code to do "HeaterOn" followed by "nextHeaterMode" when "Solar" (aka heatpump) is desired.

**Reply by bsileo** [reply]
> I use (or used) Actiontiles with smartthings. I was hoping to transition away from actiontiles and use hubitat dashboard to have all local control and not be cloud dependent, but unfortunately many options are missing from hubitat dashboard and design is difficult (in my opinion).

My experience was similar so I ended up integrating Sharptools. Inherently you need something off the lan for control when away from home (at least I do) so this checked that box and works for local control as well.

**Reply by bsileo** [reply]
> > What device are you looking at for heater control? This will be on a body, such as the Pool Body or Spa Body.
> 
> Yes, I'm looking at pool and/or spa body. From the device page as you showed, the setpoint commands and heater off command works but no response with heater on or next heater mode.
> 
> Within hubitat dashboard, the only templates that seem to work are attributes and temperature but these only show the data or attribute chosen, no command options. I've read with other pool integrations, the thermostat template can be used for the heater.

The Pool Controller Body device type includes these capabilities:
"TemperatureMeasurement"
"ThermostatHeatingSetpoint"

So you are correct that you should be able to use a Thermostat template to interact with it.  Here is an example of this on my dashboard (in Sharptools):

![image](https://user-images.githubusercontent.com/6116016/170703053-cdfe8c5c-a81d-46b5-886f-7617a3f6a02f.png)

I do not use the Hubitat Dashboards so I can't quickly test this side out. Next time I get deeper into development I can try to install those on my Dev hub and see if it works. It certainly should since this is all standard Hubitat settings.
